### ISSUE TITLE
[SAE] AdaptiveTopK.penalty() is detached — cannot train the learned k head

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#611]
**Verified — CONFIRMED.**

`forward` stores the mean detached and under no_grad (`gamfit/torch/modules.py:266-267`):
```python
with torch.no_grad():
    self._last_k_pred_mean = k_pred.detach().mean()
```
and `penalty()` returns `exp(log_weight) * self._last_k_pred_mean` (~270-272). The custom backward in `_AdaptiveTopKSTE` (~131-150) does not route reconstruction gradient through the threshold/order-statistic into `k_pred`. So `loss = recon + adaptive_topk.penalty()` cannot train the `k_head`; it's trained only by reconstruction, not by its own sparsity penalty. The advertised adaptive/learned top-k mechanism is not actually trained by the penalty.

**Fix:** keep a graph-connected mean (`self._last_k_pred_mean_graph = k_pred.mean()`) and use it in `penalty()`, or return the sparsity penalty directly from `forward()` so it corresponds to the current batch. Stronger: implement a differentiable top-k relaxation (`z_hard + z_soft - z_soft.detach()`) so gradients flow through the surrogate.

_Source: static SAE/manifold-SAE audit; verified against current code at commit 3c1a4efa2._

**Comment by SauersML** [comment]
Fixed in `12d764cec`: AdaptiveTopK keeps a graph-connected `_last_k_pred_mean_graph = k_pred.mean()` used by `penalty()`, and forward returns the sparsity penalty in its tuple, with an STE soft-surrogate (`z_hard + z_soft - z_soft.detach()`) — the learned-k head now trains from its own penalty (modules.py:291-316).
